src/Eshopworld.WorkerProcess/LeaseAllocator.cs

```diff
@@ -34,83 +35,103 @@ public class LeaseAllocator : ILeaseAllocator
+                lease = createResult.Lease;
+            }
 
-                        return updateResult.Lease;
-                    }
 
-                    _telemetry.Publish(new LeaseAcquisitionEvent(instanceId, _options.Value.WorkerType,
-                        _options.Value.Priority,
-                        $"Lease activation failed lease already acquired. Allocated Lease Id: [{lease.Id}]"));
-                }
+            lease.Priority = _options.Value.Priority;
+            lease.InstanceId = instanceId;
+            var updateTime = ServerDateTime.UtcNow;
+            lease.Interval = _slottedInterval.Calculate(updateTime, _options.Value.LeaseInterval);
+            lease.LeasedUntil = updateTime.Add(lease.Interval.Value);
+            var updateResult = await TryAcquireLease(lease, instanceId, _leaseStore.TryUpdateLeaseAsync).ConfigureAwait(false);
+            if (updateResult.Result || updateResult.Lease == null)
+            {
+                return updateResult.Lease;
+            }
+            _telemetry.Publish(new LeaseAcquisitionEvent(instanceId, _options.Value.WorkerType,
+                _options.Value.Priority,
+                $"Lease failed to be acquired on update. {GenerateLeaseInfo(updateResult.Lease)}"));
+            return null;
+           
+        }
 
-                // Read the lease for event
-                lease = await _leaseStore.ReadByLeaseTypeAsync(_options.Value.WorkerType).ConfigureAwait(false);
+        private async Task<LeaseStoreResult> TryAcquireLease(ILease lease, Guid instanceId, Func<ILease,Task<LeaseStoreResult>> leaseAcquirer)
+        {
+            var leaseResult = await leaseAcquirer(lease).ConfigureAwait(false);
 
+            if (leaseResult.Result)
+            {
                 _telemetry.Publish(new LeaseAcquisitionEvent(instanceId, _options.Value.WorkerType,
                     _options.Value.Priority,
-                    $"Lease acquisition failed lease already acquired. {GenerateLeaseInfo(lease)}"));
+                    $"Lease acquired. {GenerateLeaseInfo(leaseResult.Lease)}"));
+                return leaseResult;
             }
-            else
+
+            // Read the lease for event
+            lease = await _leaseStore.ReadByLeaseTypeAsync(_options.Value.WorkerType).ConfigureAwait(false);
+
+            if (lease.Priority <= _options.Value.Priority)
             {
-                // lease already acquired
+                // if we get here, means that there was a higher priority wp claiming to be the winner and acquiring the lease previously
+                // this happens if not all worker processes see the same winner, despite of the delay. Give up in this case
+
                 _telemetry.Publish(new LeaseAcquisitionEvent(instanceId, _options.Value.WorkerType,
                     _options.Value.Priority,
                     $"Lease acquisition failed lease already acquired. {GenerateLeaseInfo(lease)}"));
```

Comment:
acquired... by someone else? may be worth adding it to the log desc

Reply:
Absolutely. I'll try to give more info in the telemetry

Reply:
resolved as per comments below